bin/requirements.rb

```diff
@@ -0,0 +1,187 @@
+    PACKAGES.each do |package|
+      filename = "#{dirname}/#{package}"
+      if File.exist?(filename)
+        @filenames << filename
+      else
+        $stderr.puts("NOTICE: missing #{filename}")
+      end
+    end
+
+    self
+  end
+
+  def parse
+    filenames.each do |fn|
+      # the list of requirements-files can have items commented out - ignore those
+
+      mod = module_name_from_filename(fn)
+      IO.read(fn).split("\n").each do |line|
+        lib, ver = parse_line(line)
+        next unless lib
+
+        # ===
+        # === business rules
+        # ===
+
+        # skip git libraries. git>= line from vsphere gave us problems
+        next if lib.start_with?("git")
+
+        # Do not version system provided libraries: requests, requests[security], pyspnego[kerberos]
+        ver = "" if lib.match?(/^requests($|\[)/) || lib.start_with?("pyspnego")
+
+        final[lib] ||= {}
+        (final[lib][ver] ||= []) << mod
+      end
+    end
+
+    self
+  end
+
+  def output
+    result = final.flat_map do |lib, vers|
+      # consolidate multiple versioning rules
+      if vers.size > 1
+        max_key, *all_keys = vers.keys
+        all_keys.each do |alt|
+          higher, lower, conflict = version_compare(alt, max_key)
+          # There is a conflict when we have something like >=2.0 and ==1.0
+          #
+          # We are just displaying all comparisons/winners to verify comparisons (skipping when a blank is provided)
+          $stderr.puts "#{lib}: #{higher} > #{lower} #{"CONFLICT" if conflict}" if lower != ""
+          vers[higher].concat(vers.delete(lower))
+          max_key = higher
+        end
+      end
+
+      vers.map do |(ver, modules)|
+        # if we pass in a previous requirements.txt, lets not mention it
+        modules.delete("legacy") if modules.size > 1
+        "#{lib}#{ver} # #{modules.join(", ")}"
+      end
+    end.sort.join("\n")
+
+    puts result
+  end
+
+  private
+
+  def module_name_from_filename(fn)
+    if non_modules.include?(fn)
+      "legacy"
+    else
+      fn.gsub(%r{.*ansible_collections/}, "")
+        .gsub(%r{/requirements.*}, "")
+    end
+  end
+
+  def parse_line(line)
+    line.downcase!
+    # TODO: do we want to keep legacy comments? Only useful for our requirements.txt file
+    line.gsub!(/#.*/, "")
+    line.strip!
+    return if line.empty?
+
+    # some libraries list "python" instead of "python_version"
+    line.gsub!(/^python /, "python_version")
+    # change version 5+ to >=5
+    line.gsub!(/\([0-9.]*\)[+]/, '>=\1')
+    line.gsub!("= ", "=")
+    # ignore line where a module requires a minumum python version (assumption here)
+    return if line.match?(/^python_version/)
+    # ignore package requirements for older version of pythons (assumption here)
+    return if line.match?(/python_version ?[=<]/)
+
+    # split on first space (or =)
+    # version can have multiple spaces
+    lib, ver = line.match(/([^ >=]*) ?(.*)/).captures
```

Comment:
I'm not sure I understand this line, but I think this can more simply be

```suggestion
    # split on first space (or =)
    # version can have multiple spaces
    lib, ver = line.split(/[ =>]/, 2)
```

Reply:
nice. Tested with a few and it works for most everything but this breaks down:

```
library==3.4
```

It looses the first `=` and turns into `library=3.4`

Maybe we want something like:
```
"library==22 ; <3".split(/([>= ]+)/, 2)
```

I'd then rewrite the comparison routine to keep the comparator separate, though then again, it does work the way it is.